src/v/crypto/include/crypto/crypto.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "bytes/bytes.h"
+#include "crypto/exceptions.h"
+#include "crypto/types.h"
+
+#include <cstddef>
+#include <memory>
+
+namespace crypto {
+///////////////////////////////////////////////////////////////////////////////
+/// Digest operations
+///////////////////////////////////////////////////////////////////////////////
+
+/**
+ * Structure used to hold the context of digest operations
+ */
+struct digest_context {
```

Comment:
it would be cool if we can prevent users from making instances of this class since its in the public header and it would be no use to them as it stands. Maybe inheritance , private constructor or something else can make this detail transparent to the user

Reply:
Users of this library cannot.  If for example you tried to do `crypto::digest_context ctx;`, this would result in a compilation error that starts with:

```
/home/mboquard/dev/redpanda/vbuild/llvm/install/bin/../include/c++/v1/__memory/unique_ptr.h:63:19: error: invalid application of 'sizeof' to an incomplete type 'crypto::digest_context::impl'
    static_assert(sizeof(_Tp) >= 0, "cannot delete an incomplete type");
```